docs: release process (#1817)
opendp/opendp: .github/workflows/README.md

- Pushes HTML to gh-pages branch, which is linked to https://docs.opendp.org

## Knobs on the Release Workflow

One-off releases can be made with the
[`release.yml` workflow](https://github.com/opendp/opendp/actions/workflows/release.yml)
on github, or with the `gh` command line tool. Parameters:

- **Target channel** controls how the release is tagged, and what semantic version is given to the release. There is a git branch with the same name for each channel.
- The **sync the Channel from upstream?** checkbox is for when you want to update the `nightly`, `beta` `stable` or `dev` branches. This replaces the contents of the respective branch with the branch the workflow is run from ("Use workflow from"). In practice, you'll almost always enable this.
- Update the **version counter** accordingly when you want to release multiple nightlies or betas in the same day.
- **Dry runs** get sent to test-pypi, and don't update the docs
- **Fake** is for developer convenience when debugging CI: it skips compilation and inserts dummy binaries instead

## Making a release

> 1. Ensure that changelog is up-to-date and version in `main` matches the version you want to release (otherwise see step 3).

Reviewer: >  1. Ensure that changelog is up-to-date

"Ensure" sounds like we're just checking, while we really need to author the changelog at the start of this process, I think?

```diff
- version in `main`
+ the `VERSION` file on the `main` branch
```

> 2. We use a release train with nightly, beta and stable channels. Repeatedly run the Release workflow through each channel (see [Making one-off releases](#Making-one-off-releases)).

Reviewer: ```suggestion
2. We use a release train with nightly, beta and stable channels. Repeatedly run the Release workflow through each channel (see preceding section).
```

I find really short hop links a little confusing. If you do want to keep it, update the link to reflect the current title.

    1. Wait until `nightly` is updated. It is also possible to update it yourself:
        * Use workflow from: `Branch: main`
        * Target Channel: `nightly`
        * Sync the channel from Upstream?: `yes`

    1. Once `nightly` is updated, update `beta`.
        * Use workflow from: `Branch: nightly`
        * Target Channel: `beta`
        * Sync the channel from Upstream?: `yes`

    1. Once `beta` is updated, update `stable`.
        * Use workflow from: `Branch: beta`
        * Target Channel: `stable`
        * Sync the channel from Upstream?: `yes`

    If CI fails, debug on main and restart.
    If beta release failed, don't forget to increment the beta version counter next time.
    If stable release failed, consider incrementing patch version and starting over.

1. Update `main`.

    1. Update version numbers:

        ```shell
        python tools/channel_tool.py bump_version --position minor
        ```

>         Add arguments as needed to bump or set version.

Reviewer: I would like this change, incrementing the version numbers right after release, but I feel like we might have talked about it earlier and I had thought you weren't sold on it? I may be misremembering.

Right now, if I look at [pypi](https://pypi.org/project/opendp/#history) or [crates.io](https://crates.io/crates/opendp/versions?sort=semver) (sorted by semver) the highest version is v0.10.0, and all the nightlies come below. With this change to our process, there will only be one v0.11 release, and the next nightly will be under v0.12... which makes me wonder if we actually want `--position patch` instead of `--position minor`? We might not actually make a v0.11.1 release and instead go to v0.12 in a few months, but we would have v0.11.1-nightlies.

Author: I had lost this piece from my brain at some point, thanks for restoring it. Glad we have it documented now.

- Pushes HTML to gh-pages branch, which is linked to https://docs.opendp.org

## Knobs on the Release Workflow

One-off releases can be made with the
[`release.yml` workflow](https://github.com/opendp/opendp/actions/workflows/release.yml)
on github, or with the `gh` command line tool. Parameters:

- **Target channel** controls how the release is tagged, and what semantic version is given to the release. There is a git branch with the same name for each channel.

> - The **sync the Channel from upstream?** checkbox is for when you want to update the `nightly`, `beta` `stable` or `dev` branches. This replaces the contents of the respective branch with the branch the workflow is run from ("Use workflow from"). In practice, you'll almost always enable this.

Reviewer: ```suggestion
- The **sync the Channel from upstream?** checkbox is for when you want to update the `nightly`, `beta`, `stable`, or `dev` branches. This replaces the contents of the respective branch with the branch the workflow is run from ("Use workflow from"). In practice, you'll almost always enable this.
```

1. We use a release train with nightly, beta and stable channels. Repeatedly run the Release workflow through each channel (see preceding section).

    1. Wait until `nightly` is updated. It is also possible to update it yourself:
        * Use workflow from: `Branch: main`
        * Target Channel: `nightly`
        * Sync the channel from Upstream?: `yes`

    1. Once `nightly` is updated, update `beta`.
        * Use workflow from: `Branch: nightly`
        * Target Channel: `beta`
        * Sync the channel from Upstream?: `yes`

    1. Once `beta` is updated, update `stable`.
        * Use workflow from: `Branch: beta`
        * Target Channel: `stable`
        * Sync the channel from Upstream?: `yes`

    If CI fails, debug on main and restart.
    If beta release failed, don't forget to increment the beta version counter next time.
    If stable release failed, consider incrementing patch version and starting over.

1. Increment patch version in `main`.

    ```shell
    python tools/channel_tool.py bump_version --position patch
    ```

>     Add arguments as needed to bump or set version.

Reviewer: ```suggestion
```

At this point in the process, aren't we always going to only bump the patch version? The comment makes it sound like there are cases where we might do something different.

## Making a release

> 1. Check that changelog is up-to-date and the `VERSION` file on the `main` branch matches the version you want to release. If not, see step 3.

Reviewer: Maybe talk this over in person? I might be thinking about this the wrong way, but unless the plan is to start updating the changelog incrementally (which isn't impossible, but in my experience leads to merge conflicts), updating the changelog needs to be at the start of the release process, not the end.

Reviewer: The auto-number is nice, but it does mean that if a step is inserted, "3" might be pointing at the right thing... but my preference is really to just get rid of this branch in the flow chart, and this would be moot.

Reviewer: Follow-up from discussion: 
- If it's a patch release, no change to the version number should be needed: We incremented the patch after the previous release.
- If it's a minor release, `python tools/channel_tool.py bump_version --position minor`
- Similarly, if it's a major release, `python tools/channel_tool.py bump_version --position major`
